binder/src/main/java/io/grpc/binder/internal/BinderServer.java

```diff
@@ -97,8 +101,9 @@ public IBinder getHostBinder() {
 
   @Override
   public synchronized void start(ServerListener serverListener) throws IOException {
-    listener = new ActiveTransportTracker(serverListener, terminationListener);
+    listener = new ActiveTransportTracker(serverListener, this::onTerminated);
```

Comment:
Actually the API contract for InternalServer#start() says "Implementations must not call {@code listener} until after {@code start()} returns" so I think what you are worried about is already impossible. But LMK if I'm missing your point. 

I put the termination code in a new method so I could give it a meaningful name rather than an anonymous lambda.

Reply:
Thanks but I don't see the connection to that tott since this PR isn't defining/changing any interfaces. ActiveTransportTracker's constructor javadoc says that terminationListener is "invoked only once the server has started shutdown ({@link #serverShutdown()} AND the last active transport is terminated" which we know isn't true at this point. Maybe you are saying we should be defensive against an ActiveTransportTracker implementation that doesn't obey this contract? I really don't know about that -- calling terminationListener at the right time is, without exaggeration, ATT's only job.

Reply:
>I don't see the connection to that tott since this PR isn't defining/changing any interfaces

Yeah, that's why I said "a bit". I was thinking in terms of relying more on compile-time guarantees against NPEs than on everyone respecting their contracts, which is of course necessary but demands appropriate test coverage.

> Maybe you are saying we should be defensive against an ActiveTransportTracker implementation that doesn't obey this contract?

Not only that, but also now we have a private `onTerminated` function that `BinderServer` itself may call inadvertently at the wrong time. Imagine if the next PR adds more initialization to `start` that may fail and call `onTerminated`:

```java
  @Override
  public synchronized void start(ServerListener serverListener) throws IOException {
    listener = new ActiveTransportTracker(serverListener, this::onTerminated);
    try {
      initializeSomethingFallible();
    } catch (Exception e) {
      // it failed, let's cleanup everything
      this.shutdown();  // will raise NPE when recycling the executor, but the call is far away and hard to remember
    }
    this.executor = ...;
  }
```

If we use a lambda, the NPE can't happen. This pattern generally encouraged by [go/nullness](http://go/nullness) or Kotlin, where you always have to deal with nullables.

I don't feel super strongly about it since we do have an API contract and all these classes are internal, so YMMV.